Getting segmentation fault error when creating PVC automatically

**Description:**
When creating PVC automatically, if the user did not specify the storage class name operator gives segmentation fault issue like below.

```sh
E0829 17:39:49.811061       1 runtime.go:69] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
src/github.com/siddhi-io/siddhi-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:76
src/github.com/siddhi-io/siddhi-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
src/github.com/siddhi-io/siddhi-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/bin/go/src/runtime/panic.go:522
/usr/local/bin/go/src/runtime/panic.go:82
/usr/local/bin/go/src/runtime/signal_unix.go:390
src/github.com/siddhi-io/siddhi-operator/pkg/controller/siddhiprocess/deploymanager/application.go:111
src/github.com/siddhi-io/siddhi-operator/pkg/controller/siddhiprocess/siddhicontroller/siddhicontroller.go:138
src/github.com/siddhi-io/siddhi-operator/pkg/controller/siddhiprocess/siddhiprocess_controller.go:249
src/github.com/siddhi-io/siddhi-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215
src/github.com/siddhi-io/siddhi-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158
src/github.com/siddhi-io/siddhi-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
src/github.com/siddhi-io/siddhi-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
src/github.com/siddhi-io/siddhi-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/usr/local/bin/go/src/runtime/asm_amd64.s:1337
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x10192b3]

goroutine 351 [running]:
github.com/siddhi-io/siddhi-operator/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	src/github.com/siddhi-io/siddhi-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x105
panic(0x127dda0, 0x21c4b20)
	/usr/local/bin/go/src/runtime/panic.go:522 +0x1b5
github.com/siddhi-io/siddhi-operator/pkg/controller/siddhiprocess/deploymanager.(*DeployManager).Deploy(0xc0009e7928, 0xc000a9ca89, 0x7, 0xc000d24440, 0x13)
	src/github.com/siddhi-io/siddhi-operator/pkg/controller/siddhiprocess/deploymanager/application.go:116 +0x1f43
github.com/siddhi-io/siddhi-operator/pkg/controller/siddhiprocess/siddhicontroller.(*SiddhiController).CreateArtifacts(0xc0009e7bc0, 0xc000317110, 0x2, 0x2)
	src/github.com/siddhi-io/siddhi-operator/pkg/controller/siddhiprocess/siddhicontroller/siddhicontroller.go:138 +0x324
github.com/siddhi-io/siddhi-operator/pkg/controller/siddhiprocess.(*ReconcileSiddhiProcess).Reconcile(0xc0004acf00, 0xc0005444c8, 0x7, 0xc0005da920, 0x11, 0x21d8c80, 0x8, 0x200, 0x40)
	src/github.com/siddhi-io/siddhi-operator/pkg/controller/siddhiprocess/siddhiprocess_controller.go:249 +0x87d
github.com/siddhi-io/siddhi-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00029c5a0, 0x0)
	src/github.com/siddhi-io/siddhi-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215 +0x1cc
github.com/siddhi-io/siddhi-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1()
	src/github.com/siddhi-io/siddhi-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158 +0x36
github.com/siddhi-io/siddhi-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000482410)
	src/github.com/siddhi-io/siddhi-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54
github.com/siddhi-io/siddhi-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000482410, 0x3b9aca00, 0x0, 0x1, 0xc000772480)
	src/github.com/siddhi-io/siddhi-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xf8
github.com/siddhi-io/siddhi-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc000482410, 0x3b9aca00, 0xc000772480)
	src/github.com/siddhi-io/siddhi-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by github.com/siddhi-io/siddhi-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start
	src/github.com/siddhi-io/siddhi-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:157 +0x311
```

**Suggested Labels:**
Type fix

**Affected Product Version:**
0.2.0-m2   

**Steps to reproduce:**
- Create a stateful distributed app without PVC spec's storage class.

--- a/pkg/controller/siddhiprocess/artifact/artifacts.go
+++ b/pkg/controller/siddhiprocess/artifact/artifacts.go
@@ -33,7 +33,6 @@ import (
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
-	"k8s.io/apimachinery/pkg/api/resource"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/intstr"
@@ -229,9 +228,7 @@ func (k *KubeClient) CreateNATS(namespace string) error {
 func (k *KubeClient) CreateOrUpdatePVC(
 	name string,
 	namespace string,
-	accessModes []corev1.PersistentVolumeAccessMode,
-	storage resource.Quantity,
-	storageClassName string,
+	pvcSpec corev1.PersistentVolumeClaimSpec,
 	owner metav1.Object,
 ) error {
 	pvc := &corev1.PersistentVolumeClaim{}
@@ -241,19 +238,21 @@ func (k *KubeClient) CreateOrUpdatePVC(
 		pvc,
 	)
 	if err != nil && apierrors.IsNotFound(err) {
-		accessGranted := false
-		for _, accessMode := range accessModes {
-			if accessMode == corev1.ReadWriteOnce {
-				accessGranted = true
-				break
+		if pvcSpec.AccessModes != nil {
+			accessGranted := false
+			for _, accessMode := range pvcSpec.AccessModes {
+				if accessMode == corev1.ReadWriteOnce {
+					accessGranted = true
+					break
+				}
+				if accessMode == corev1.ReadWriteMany {
+					accessGranted = true
+					break
+				}
 			}
-			if accessMode == corev1.ReadWriteMany {
-				accessGranted = true
-				break
+			if !accessGranted {
+				return errors.New("Restricted access mode in " + name)
 			}
-		}
-		if !accessGranted {
-			return errors.New("Restricted access mode in " + name)
 		}
 		pvc = &corev1.PersistentVolumeClaim{
 			TypeMeta: metav1.TypeMeta{
@@ -264,22 +263,14 @@ func (k *KubeClient) CreateOrUpdatePVC(
 				Name:      name,
 				Namespace: namespace,
 			},
-			Spec: corev1.PersistentVolumeClaimSpec{
-				AccessModes: accessModes,
-				Resources: corev1.ResourceRequirements{
-					Requests: corev1.ResourceList{
-						corev1.ResourceStorage: storage,
-					},
-				},
-				StorageClassName: &storageClassName,
-			},
+			Spec: pvcSpec,
 		}
 		controllerutil.SetControllerReference(owner, pvc, k.Scheme)
 		_, err = controllerutil.CreateOrUpdate(
 			context.TODO(),
 			k.Client,
 			pvc,
-			PVCMutateFunc(accessModes, storage, storageClassName),
+			PVCMutateFunc(pvcSpec),
 		)
 	}
 	return err
@@ -529,17 +520,11 @@ func ConfigMapMutateFunc(data map[string]string) controllerutil.MutateFn {
 
 // PVCMutateFunc is the function for update k8s persistence volumes claims gracefully
 func PVCMutateFunc(
-	accessModes []corev1.PersistentVolumeAccessMode,
-	storage resource.Quantity,
-	class string,
+	pvcSpec corev1.PersistentVolumeClaimSpec,
 ) controllerutil.MutateFn {
 	return func(obj runtime.Object) error {
 		pvc := obj.(*corev1.PersistentVolumeClaim)
-		pvc.Spec.AccessModes = accessModes
-		pvc.Spec.Resources.Requests = corev1.ResourceList{
-			corev1.ResourceStorage: storage,
-		}
-		pvc.Spec.StorageClassName = &class
+		pvc.Spec = pvcSpec
 		return nil
 	}
 }

--- a/pkg/controller/siddhiprocess/artifact/artifacts_test.go
+++ b/pkg/controller/siddhiprocess/artifact/artifacts_test.go
@@ -155,9 +155,17 @@ func TestCreateOrUpdatePVC(t *testing.T) {
 		corev1.ReadWriteOnce,
 		corev1.ReadWriteMany,
 	}
-	storage := resource.MustParse("1Gi")
 	storageClassName := "standard"
-	err := kubeClient.CreateOrUpdatePVC(name, namespace, accessModes, storage, storageClassName, sampleDeployment)
+	pvcSpec := corev1.PersistentVolumeClaimSpec{
+		AccessModes:      accessModes,
+		StorageClassName: &storageClassName,
+		Resources: corev1.ResourceRequirements{
+			Requests: corev1.ResourceList{
+				corev1.ResourceStorage: *resource.NewQuantity(1*1024*1024*1024, resource.BinarySI),
+			},
+		},
+	}
+	err := kubeClient.CreateOrUpdatePVC(name, namespace, pvcSpec, sampleDeployment)
 	if err != nil {
 		t.Error(err)
 	}

--- a/pkg/controller/siddhiprocess/deploymanager/application.go
+++ b/pkg/controller/siddhiprocess/deploymanager/application.go
@@ -111,9 +111,7 @@ func (d *DeployManager) Deploy() (operationResult controllerutil.OperationResult
 			err = d.KubeClient.CreateOrUpdatePVC(
 				pvcName,
 				d.SiddhiProcess.Namespace,
-				d.SiddhiProcess.Spec.PVC.AccessModes,
-				d.SiddhiProcess.Spec.PVC.Resources.Requests[corev1.ResourceStorage],
-				*d.SiddhiProcess.Spec.PVC.StorageClassName,
+				d.SiddhiProcess.Spec.PVC,
 				d.SiddhiProcess,
 			)
 			if err != nil {

